driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java

```diff
@@ -228,187 +184,356 @@ public int getMaxWireVersion() {
-                        : t;
-                connection.release();
-                endOperationInProgress();
-                callback.onResult(null, translatedException);
+        void setServerCursor(@Nullable final ServerCursor serverCursor) {
+            assertTrue(state.inProgress());
+            assertNotNull(connectionSource);
+            this.serverCursor = serverCursor;
+            if (serverCursor == null) {
+                releaseClientResources();
+            }
+        }
+
+        private void releaseServerAndClientResources(final AsyncConnection connection) {
+            lock.lock();
+            ServerCursor localServerCursor = serverCursor;
+            serverCursor = null;
+            lock.unlock();
+            if (localServerCursor != null) {
+                killServerCursor(namespace, localServerCursor, connection);
             } else {
-                assertNotNull(result);
-                handleGetMoreQueryResult(connection, callback,
-                        initFromCommandCursorDocument(connection.getDescription().getServerAddress(), NEXT_BATCH, result));
+                releaseClientResources();
+            }
+        }
+
+        private BsonDocument asKillCursorsCommandDocument(final MongoNamespace namespace, final ServerCursor serverCursor) {
+            return new BsonDocument("killCursors", new BsonString(namespace.getCollectionName()))
+                    .append("cursors", new BsonArray(singletonList(new BsonInt64(serverCursor.getId()))));
+        }
+
+        private void releaseServerResources(final AsyncConnection connection) {
+            ServerCursor localServerCursor = serverCursor;
+            serverCursor = null;
+            if (localServerCursor != null) {
+                killServerCursor(namespace, localServerCursor, connection);
+            }
+        }
+
+        private void killServerCursor(final MongoNamespace namespace, final ServerCursor serverCursor, final AsyncConnection connection) {
+            connection.retain();
+            AsyncConnectionSource connectionSource = assertNotNull(resourceManager.connectionSource).retain();
+            connection.commandAsync(namespace.getDatabaseName(), asKillCursorsCommandDocument(namespace, serverCursor),
+                    NO_OP_FIELD_NAME_VALIDATOR, ReadPreference.primary(), new BsonDocumentCodec(),
+                    connectionSource.getOperationContext(), (r, t) -> {
+                        connectionSource.release();
+                        connection.release();
+                        releaseClientResources();
+                    });
+        }
+
+        private void releaseClientResources() {
+            assertNull(serverCursor);
+            lock.lock();
+            AsyncConnectionSource localConnectionSource = connectionSource;
+            connectionSource = null;
+
+            AsyncConnection localPinnedConnection = pinnedConnection;
+            pinnedConnection = null;
+            lock.unlock();
```

Comment:
Why is this locking needed, given that `CommandBatchCursor` does not have it?

The reason `CommandBatchCursor` does not have it is because of the guarantee provided by the `ResourceManager`, and as long as that guarantee is not violated in `AsyncCommandBatchCursor` locking should not be needed neither here nor in `releaseServerAndClientResources`.

I do have doubts that that guarantee is preserved in the asynchronous code, given VAKOTODO, VAKOTODO. But if it is violated, then the rest of the code that relies on it may be incorrect, and `ResourceManager` becomes useless.

Reply:
Ack - will fix in #401